### PR TITLE
Preserve VP9 sequence numbers, picId, and tl0picIdx across suspensions

### PR DESCRIPTION
**Pull request by JonathanLennox**
Otherwise Chrome's decoder can get confused, and decode the video incorrectly.

### CODE REVIEW

File: jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt [hunk 1]
```diff
@@ -1010,13 +1010,172 @@ class Vp9AdaptiveSourceProjectionTest {
...
+            } else {
+                Assert.assertFalse(accepted)
+            }
+            if (endOfPicture) {
+                expectedTs = RtpUtils.applyTimestampDelta(expectedTs, 3000)
+                expectedPicId = applyExtendedPictureIdDelta(expectedPicId, 1)
+            }
+        }
+        var suspended = 64
+        while (suspended < 65536) {
```
Comment (bgrozev) [inline jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt:1071]: This loop is equivalent to `for(suspended = 64; suspended < 65536; suspended *= 2)`, right? I find the `for` version easier to read.

File: jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt [hunk 2]
```diff
@@ -1010,13 +1010,172 @@ class Vp9AdaptiveSourceProjectionTest {
...
+                }
+            }
+            suspended *= 2
+        }
+    }
+
+    @Test
+    fun sourceSuspensionTest() {
+        val generator = ScalableVp9PacketGenerator(1)
+        runSourceSuspensionTest(generator, 2)
```
Comment (bgrozev) [inline jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt:1177]: nit: I think `getIndex(0, 0, 2)` would be easier to understand (or `getIndex(tid=2)` if you were to add default values to `getIndex`).

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionContext.kt [hunk 3]
```diff
@@ -334,6 +335,49 @@ class Vp9AdaptiveSourceProjectionContext(
...
+            lastVp9FrameProjection.timestamp,
+            lastProjectedFrame.timestamp
+        )
+        val projectedTs = applyTimestampDelta(frame.timestamp, tsDelta)
+
+        /* Increment picId and tl0picidx by 1 from the last projected frame. */
+        val projectedPicId = applyExtendedPictureIdDelta(lastVp9FrameProjection.pictureId, 1)
+        val projectedTl0PicIdx = applyTl0PicIdxDelta(lastVp9FrameProjection.tl0PICIDX, 1)
+
+        /* Increment sequence numbers based on the last projected frame, but leave a gap
```
Comment (fippo) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionContext.kt:366]: how does that work? You can see if the current packet is not the beginning of the frame (i.e. B is not set), but you can't know how many packets of the new frame are missing?

I'd expect more than one missing/reordered packet to be quite rare though.

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9QualityFilter.kt [hunk 4]
```diff
@@ -110,7 +111,9 @@ internal class Vp9QualityFilter(parentLogger: Logger) {
                I think this shouldn't be a problem? */
             getSidFromIndex(incomingIndex) == getSidFromIndex(externalTargetIndex)
         }
-        return AcceptResult(accept, mark)
+        val isResumption = (prevIndex == SUSPENDED_INDEX && currentIndex != SUSPENDED_INDEX)
+        if (isResumption) assert(accept) // Pretty sure this has to be true?
```
Comment (fippo) [inline jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9QualityFilter.kt:115]: nit: remove the comment
I don't see how you can have a resumption and then don't accept the frame either.

**Review by fippo — APPROVED** [review]
LGTM (modulo my missing kotlin skills) with one question and a nit

File: jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionContext.kt (hunk 3 above)
Comment (JonathanLennox) [reply jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionContext.kt:366]: I'm keeping track of all received frames, not just those in frames that I've forwarded.  Thus I can calculate the sequence number gap between the most recent packet previously received and this packet, and so leave the equivalent gap in the outgoing sequence number space.

**Comment by bgrozev** [comment]
LGTM too (except for the nits).

File: jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt (hunk 1 above)
Comment (JonathanLennox) [reply jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt:1071]: Kotlin doesn't actually have C-style `for` loops, only `foreach`-style loops.

I can use a helper function of the type seen at https://stackoverflow.com/questions/66377331/how-do-you-iterate-a-for-loop-with-multiplication-or-division-kotlin .

Comment (bgrozev) [reply jvb/src/test/kotlin/org/jitsi/videobridge/cc/vp9/Vp9AdaptiveSourceProjectionTest.kt:1071]: That's annoying. You could also do something like
```
for (i in 5..16) {
  val suspended = 2.0.pow(i).toInt()
}
```

I'm fine with either version, including the original.

**Review by bgrozev — DISMISSED** [review]

**Review by bgrozev — APPROVED** [review]

**Pull request merged**